### ISSUE TITLE
Euclidean distance over factors (`match_on.formula`)

### ISSUE THREAD

**Issue by markmfredrickson** [markmfredrickson/optmatch#80]
The most logical Euclidean distance calculation over factors would be to have a zero for all pairs with the same value of the factor and some positive value otherwise. Right now, this is not happening:

```
> z <- rep(c(T,F), 4)
> f <- as.factor(rep(1:4, each = 2))
> match_on(z ~ f, method = "euclidean")
An object of class "DenseMatrix"
         control
treatment 2        4        6        8
        1 0 1.000000 1.000000 1.000000
        3 1 0.000000 1.414214 1.414214
        5 1 1.414214 0.000000 1.414214
        7 1 1.414214 1.414214 0.000000
Slot "call":
match_on(x = z ~ f, method = "euclidean")

> match_on(z ~ f + 0, method = "euclidean")
An object of class "DenseMatrix"
         control
treatment 2        4        6        8
        1 0 1.000000 1.000000 1.000000
        3 1 0.000000 1.414214 1.414214
        5 1 1.414214 0.000000 1.414214
        7 1 1.414214 1.414214 0.000000
Slot "call":
match_on(x = z ~ f + 0, method = "euclidean")

> match_on(z ~ f - 1, method = "euclidean")
An object of class "DenseMatrix"
         control
treatment 2        4        6        8
        1 0 1.000000 1.000000 1.000000
        3 1 0.000000 1.414214 1.414214
        5 1 1.414214 0.000000 1.414214
        7 1 1.414214 1.414214 0.000000
Slot "call":
match_on(x = z ~ f - 1, method = "euclidean")
```
This behavior can be approximated with

```
 match_on(z ~ ., data = data.frame(z = z, model.matrix(~ f - 1)), method = "euclid")
An object of class "DenseMatrix"
         control
treatment        2        4        6        8
        1 0.000000 1.414214 1.414214 1.414214
        3 1.414214 0.000000 1.414214 1.414214
        5 1.414214 1.414214 0.000000 1.414214
        7 1.414214 1.414214 1.414214 0.000000
```
This problem probably also affects Mahalanobis distance.

**Comment by benthestatistician** [comment]
When `model.matrix` is called inside of `match_on.formula`, it expands factors using whatever applicable function it finds in the `options("contrasts")`.  For me this is

````{r}
> options("contrasts")
$contrasts
        unordered           ordered 
"contr.treatment"      "contr.poly" 
````

So in the example Mark gives, it's `contr.treatment` that's being applied, with a result as follows:
````{r}
> (data <- subset(model.matrix(x, mf), T, -1) )
  f2 f3 f4
1  0  0  0
2  0  0  0
3  1  0  0
4  1  0  0
5  0  1  0
6  0  1  0
7  0  0  1
8  0  0  1
````

We'd like a 4 column dummy encoding here, not this 3 column version. In principle it should solve the problem to convince it to call `contr.treatment`, or whichever other contrast function it finds when it looks, with `contrasts=FALSE`, rather than the default `contrasts=TRUE`.  However, (a) it's not so clear how to do that neatly; and more important (b) documentation notwithstanding, `contr.treatment(x, contrasts=F)` appears to spit out a 3 column encoding of the factor, not a 4 column one.  One solution would be to grow our own alternative to `contr.treatment` and apply it instead.  

I'm not sure what the right behavior for ordered factors should be, but the `contr.poly` default seems to give the desired behavior in Mark's example. which suggests that maybe the way to go is to set the default "contrasts" option to "contr.poly" for both ordered and unordered factors, at least within this function.

**Comment by benthestatistician** [comment]
Update: `contr.poly` appears to give good answers with multiple factors, too. 
```
> options(contrasts=rep("contr.poly",2))
> z <- rep(c(T,F), 4)
> f1 = as.factor(rep(rep(1:2, each=2),2)) ; f2 = as.factor(rep(3:4, each=4))
> data.frame(z,f1,f2)
      z f1 f2
1  TRUE  1  3
2 FALSE  1  3
3  TRUE  2  3
4 FALSE  2  3
5  TRUE  1  4
6 FALSE  1  4
7  TRUE  2  4
8 FALSE  2  4
> match_on(z ~ f1+f2, method = "euclidean")
An object of class "DenseMatrix"
         control
treatment        2        4        6        8
        1 0.000000 1.414214 1.414214 2.000000
        3 1.414214 0.000000 2.000000 1.414214
        5 1.414214 2.000000 0.000000 1.414214
        7 2.000000 1.414214 1.414214 0.000000
Slot "call":
match_on(x = z ~ f1 + f2, method = "euclidean")

````

Thoughts on how to make a special value of `options(contrasts)` hold inside of `match_on.formula`, anyone?

**Comment by benthestatistician** [comment]
The `contr.poly` solution gives non-consonant answers across different representations of similar information.
````
> options(contrasts=rep("contr.poly",2))
> match_on(z~as.numeric(f1), method = "euclidean")
An object of class "DenseMatrix"
         control
treatment 2 4 6 8
        1 0 1 0 1
        3 1 0 1 0
        5 0 1 0 1
        7 1 0 1 0
Slot "call":
match_on(x = z ~ as.numeric(f1), method = "euclidean")

>           match_on(z~f1, method = "euclidean") 
An object of class "DenseMatrix"
         control
treatment        2        4        6        8
        1 0.000000 1.414214 0.000000 1.414214
        3 1.414214 0.000000 1.414214 0.000000
        5 0.000000 1.414214 0.000000 1.414214
        7 1.414214 0.000000 1.414214 0.000000
Slot "call":
match_on(x = z ~ f1, method = "euclidean")
````

A solution is to use a variant of `contr.Poly` that's scaled by $2^{-1/2}$.  I found that `contr.poly` is being called in a way that differs from what's documented -- among other things, changing its `scores` argument doesn't seem to do much -- but I found that something like this
````
contr.Poly <- function(n, scores=NULL, contrasts=TRUE, sparse=FALSE)
    {
        if (is.null(scores))
            scores <- if (length(n)>1) 1:length(n) else 1:as.numeric(n)
        
        contr.poly(n, scores, contrasts, sparse)/sqrt(2)
    }
options(contrasts=rep("contr.Poly",2))
````
causes each of the following to produce the same 0/1 matrix, the one that only `z~as.numeric(f1)` generates when contrasts are as given by `contr.poly`:
````
match_on(z~f1, method="euclidean")
match_on(z~I(f1=="1"), method="euclidean")
match_on(z~as.numeric(f1), method="euclidean") 
````
This seems a preferable solution to unmodified `contr.poly`.

By the way, still interested in any advice on getting `match_on.formula` to temporarily override whatever value of `options(contrasts)` happens to be in effect.  (I have the nagging feeling that there's a best practice on this that I'm not aware of.)

**Comment by benthestatistician** [comment]
Pushed a new branch, issue80, containing one solution to this problem.  The solution involves defining a new function `contr.match_on`.  Originally intended to make it a private function, but that didn't work; I had to export it to get it to be visible when needed.  If you can think of a way to export this but avoid having an entry for it in the manual, let me know.

**Comment by markmfredrickson** [comment]
I don't know if it is possible to have `contr.match_on` be a local function and use either a) `options` or b) `contrasts(column) <- "contr.match_on"`. Both of these go looking for the function (by name) in the `parent.env()`. It does not seem that `"optmatch:::contr.match_on` is considered valid.

Is there a reason you don't like manually setting the `contrasts.arg` argument in `model.matrix`? I see you commented out a solution that used this approach.

Pending a decision on this final concern, I think this is ready to go into master.

**Comment by benthestatistician** [comment]
Part of the problem with the commented-out material was that it hard-codes use of `contr.match_on` as opposed to some other contrast function the user might specify by way of `options`.
```
###  isF <- vapply(mf, is.factor, TRUE)
###  c.arg <- as.list(rep("contr.match_on", sum(isF)))
###  names(c.arg) <- names(mf)[isF]
###  data <- subset(model.matrix(x, mf, contrasts.arg=c.arg), T, -1)  
```
Looking at it with fresh eyes, however, overriding the contrasts option may be a feature rather than a bug.  If the user sets a global option for contrasts, presumably her intention is to influence the way factors figure as independent variables in regression models, not to influence the workings of `match_on.glm`.   To offer the user a route, if not that route, to controlling the contrasts, we could do something like
```
for (thefactor in which(isF))
{
if (!is.null(attr(mf[[thefactor]], "contrasts"))) contrasts(mf[[thefactor]]) <- "contr.match_on"
}
```

Unfortunately I can't recall my second concern with the commented-out approach, the one about "data duplication" or "factor duplication".  Maybe I thought there'd be a memory advantage to the option-setting approach (although I'm not seeing that now).  It's also possible that the commented-out code generated incorrect results; if so, then I'm sure it's something one could see in straightforward variations on the examples appearing in the tests (otherwise I would have added a test). 

In short, I'm happy to have it be reverted to the explicit contrast-setting version, if that could be part of a workaround to avoid having to export `contr.match_on`, such as maybe defining `contr.match_on` only locally within `match_on.glm`.

**Comment by markmfredrickson** [comment]
Made the function private by calling it directly (to generate matrices) instead of having `model.matrix` call it.
